// jsbin/tmp/store.js
var templates = require('./sql_templates'),
    utils = require('../utils');

function getArgs() {
  var args = [].slice.apply(arguments);
  if (typeof args[args.length - 1] !== 'function') {
    args.push(function () {});
  }

  return {
    query: args.shift(),
    callback: args.pop(),
    values: args
  };
}

module.exports = utils.inherit(Object, {
  defaults: null,
  connection: null,
  options: null,
  constructor: function Store(options) {
    this.options = options;
    this.defaults = {html: '', css: '', javascript: ''};
  },
  connect: function (callback) {

  },
  disconnect: function (callback) {

  },
  query: function () {
    var args = getArgs.apply(this, arguments);
    this.connection.query(args.query, args.values, args.callback);
  },
  insert: function () {
    var args = getArgs.apply(this, arguments);
    this.connection.insert(args.query, args.values, args.callback);
  },
  update: function () {
    var args = getArgs.apply(this, arguments);
    this.connection.update(args.query, args.values, args.callback);
  }
});



/*

// getBin
Bin.findOne({ url: url, revision }, fn);
// getLatestBin
Bin.find({ url: url }, { sort:  })


 */
